cpp/src/io/parquet/experimental/hybrid_scan_impl.cpp

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright (c) 2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "hybrid_scan_impl.hpp"
+
+#include "cudf/io/text/byte_range_info.hpp"
+#include "hybrid_scan_helpers.hpp"
+
+#include <cudf/detail/stream_compaction.hpp>
+#include <cudf/detail/transform.hpp>
+#include <cudf/detail/utilities/stream_pool.hpp>
+#include <cudf/io/parquet_schema.hpp>
+#include <cudf/strings/detail/utilities.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/utilities/memory_resource.hpp>
+
+#include <thrust/iterator/counting_iterator.h>
+
+#include <bitset>
+#include <iterator>
+#include <limits>
+#include <numeric>
+
+namespace cudf::io::parquet::experimental::detail {
+
+using ColumnChunkDesc       = parquet::detail::ColumnChunkDesc;
+using decode_kernel_mask    = parquet::detail::decode_kernel_mask;
+using PageInfo              = parquet::detail::PageInfo;
+using PageNestingDecodeInfo = parquet::detail::PageNestingDecodeInfo;
+using byte_range_info       = text::byte_range_info;
+
+impl::impl(cudf::host_span<uint8_t const> footer_bytes,
+           cudf::io::parquet_reader_options const& options)
+{
+  // Open and parse the source dataset metadata
+  _metadata = std::make_unique<aggregate_reader_metadata>(
+    footer_bytes,
+    options.is_enabled_use_arrow_schema(),
+    options.get_columns().has_value() and options.is_enabled_allow_mismatched_pq_schemas());
+}
+
+FileMetaData const& impl::get_parquet_metadata() const { return _metadata->get_parquet_metadata(); }
+
+byte_range_info impl::get_page_index_bytes() const { return _metadata->get_page_index_bytes(); }
+
+void impl::setup_page_index(cudf::host_span<uint8_t const> page_index_bytes) const
+{
+  _metadata->setup_page_index(page_index_bytes);
+}
+
+std::vector<size_type> impl::get_all_row_groups(
+  cudf::io::parquet_reader_options const& options) const
+{
+  auto const num_row_groups = _metadata->get_num_row_groups();
+  auto row_groups_indices   = std::vector<size_type>(num_row_groups);
+  std::iota(row_groups_indices.begin(), row_groups_indices.end(), size_type{0});
+  return row_groups_indices;
+}
+
+std::vector<std::vector<size_type>> impl::filter_row_groups_with_stats(
```

Comment:
Only contains empty definitions for impl's member functions declared in reader_impl.hpp

Reply:
I assume that these functions are going to all be filled in in a future PR?

Reply:
Yep, draft PRs are already open for some of these, just waiting for this PR to merge.